clamd/Makefile.am

```diff
@@ -46,6 +46,26 @@ if INSTALL_SYSTEMD_UNITS
 systemdsystemunit_DATA = clamav-daemon.socket clamav-daemon.service
 endif
 
+if INSTALL_OPENRC_SERVICES
+
+openrcdir = $(sysconfdir)/init.d
+
+# The next rule allows us to replace @RUNSTATEDIR@ and friends within
+# our OpenRC service scripts. This example is taken from the autoconf
+# documentation in the "Installation Directory Variables" section.
+openrc_SCRIPTS_src = openrc/clamd.in.in
+openrc_SCRIPTS_intermediate = openrc/clamd.in
```

Comment:
The two-stage substitution and the capital variable names work around a problem with the autoconf `AC_CONFIG_FILES` substitution. If you try to put something like `FRESHCLAM_PATH="@bindir@/freshclam"` in `foo.in`, and then put `foo.in` in `AC_CONFIG_FILES`, you don't get `FRESHCLAM_PATH="/usr/bin/freshclam"` as its replacement, but instead you get literally `FRESHCLAM_PATH="${prefix}/bin/freshclam"`. If `foo` is a script and you try to run it, the variable `${prefix}` is going to be undefined, and `FRESHCLAM_PATH` will be `/bin/freshclam`. Not what you want.

By using a capital `@BINDIR@`, we prevent autoconf from making that substitution, thereby leaving it for the "edit" rule in the makefile. If we don't need any autoconf substitutions _at all_, then yeah, we could get away with just one level of substitution, where the "edit" rule in the makefile turns `foo.in` to `foo`. But then you have an `.in` file that works weird. Most other `.in` files have _all_ autoconf variables substituted into them, and these would only have the special-case substitutions from the "edit" rule. What we would really like is to have both, the usual autoconf substitutions, and the `@BINDIR@` et alia substitutions that sidestep the `${prefix}` problem.

To motivate that claim a little bit, notice how I've left comments to the effect of "Foreground: no must be set in clamd.conf." It would be nice if we (me, Gentoo) did that for our users, maybe by running a pre-flight grep on the config file in the clamd OpenRC script. Well if we're using grep, then I would call `AC_PROG_GREP` to check for it and to get its path, and then instead of running `grep <whatever>` in the script, I would use `@GREP@ <whatever>` so that autoconf substitutes the correct absolute path in there. And then the OpenRC script would have both types of substitutions, a pure-autoconf one, and the "custom" one for `@RUNSTATEDIR@`. Two levels of `.in` because the two levels of substitution get performed sequentially.

There aren't any files that need that right now, but the OpenRC scripts are likely to in the future. I can also imagine it for the config files, if you want to set something like the PCRE recursion example setting to the system default.

I added the extra level of indirection now so that they all work how you'd expect after this commit, even if the `.in.in` is a bit ugly. The alternative would be to get away with a single `.in` for a while, and then to have to go through all this mess later, the first time you need an autoconf substitution in one of those files.

tl;dr:

Autoconf's `AC_CONFIG_FILES` performs most variable substitutions, but it messes up `@bindir@` and friends. The "edit" rule in the makefile handles `@BINDIR@` and friends, but that's it. To get the best of both worlds, we need both.